### ISSUE TITLE
Cache issue when checking modified files

### ISSUE THREAD

**Issue by Vitamin4iK-afk** [vimeo/psalm#11093]
I encountered a problem with Psalm cache when checking modified files.

During code inspection Psalm reported error: `ERROR: PossiblyInvalidMethodCall - file:line:position - Cannot call method on possible array<array-key, int> variable (see https://psalm.dev/113)`. I fixed this error and re-run code inspection, but Psalm still reports this error. Only when I cleared cache with `psalm --clear-cache` the error stopped appearing on further inspections.

**Steps to Reproduce:**
I managed to make a minimal repository to reproduce this problem: https://github.com/Vitamin4iK-afk/psalm-test.
1. Checkout initial commit.
2. Clear Psalm cache with psalm --clear-cache.
3. Run code inspection. This correctly reports error.
4. Checkout latest commit.
5. Run code inspection again. This falsely reports already fixed error.

**System Information:**
Operating System: Ubuntu 22.04.4 LTS
PHP Version: 8.2.22, 8.3.11
Psalm Version: 5.25.0

**Additional Information:**
Clearing Psalm cache resolves the problem but significantly increases inspection time and is not suitable for large repositories.

**Comment by Vitamin4iK-afk** [comment]
> Hey @Vitamin4iK-afk, can you reproduce the issue on https://psalm.dev? These will be used as phpunit tests when implementing the feature or fixing this bug.

No. This issue can`t be reproduced on https://psalm.dev/ because to reproduce it you need to follow certain steps described in the issue.

**Comment by weirdan** [comment]
Reproduced as described.

**Comment by theodorejb** [comment]
I encounter this issue so frequently when switching between branches/commits, that I can rarely run Psalm successfully without manually clearing the cache.

**Comment by Robert-Bartel** [comment]
We have the same issue. Exactly as described above.

Is there any solution without clearing the cache?

**Comment by fenKss** [comment]
Faced the same problem. I'm looking forward to being corrected, I'm tired of constantly watching her, because checks usually take far from a couple of seconds

**Comment by Tory495** [comment]
Fix that plz

**Comment by M-E-M-F-I-S** [comment]
Encountered same issue on our project, is this planned to be fixed?

**Comment by danog** [comment]
I would recommend using `--no-cache` with JIT enabled for the time being, will probably work more on cache in the coming months.

**Comment by jcvignoli** [comment]
6 months later, another small encouragement to work on the cache. It's unreliable.

**Comment by danog** [comment]
https://github.com/vimeo/psalm/pull/11416 will hotfix most issues with the cache, which I believe are actually related to diff mode being automatically enabled when cache is enabled.

**Comment by danog** [comment]
Merged.
